File: TableViewAgent/ViewObject/FRCAgentViewObject.m
//
// Created by P.I.akura on 2013/09/27.
// Copyright (c) 2013 P.I.akura. All rights reserved.
//
// To change the template use AppCode | Preferences | File Templates.
//


#import "FRCAgentViewObject.h"
#import "TableViewAgentCategory.h"

@interface FRCAgentViewObject () <NSFetchedResultsControllerDelegate>
@property (nonatomic) NSFetchedResultsController * controller;
@end

@implementation FRCAgentViewObject

- (id)initWithFetch:(NSFetchedResultsController *)controller {
    self = [super init];
    if (self) {
        self.controller = controller;
        self.controller.delegate = self;
    }
    return self;
}
- (void)dealloc {
    _controller = nil;
}
- (NSUInteger)sectionCount {
    if (_controller.sections.count == 1 && [_controller.sections[0] numberOfObjects] == 0) {
        return 0;
    }
    return _controller.sections.count;
}
- (NSUInteger)countInSection:(NSUInteger)section {
    return [_controller.sections[section] numberOfObjects];
}
- (id)objectAtIndexPath:(NSIndexPath *)indexPath {
    return [_controller objectAtIndexPath:indexPath];
}
- (void)removeObjectAtIndexPath:(NSIndexPath *)indexPath {
}
- (BOOL)existObject:(NSIndexPath *)indexPath {
    return indexPath.section < [self sectionCount] && indexPath.row < [self countInSection:indexPath.section];
}
- (NSArray *)sectionObjects:(NSInteger)section {
    return @[[self objectAtIndexPath:[NSIndexPath indexPathForRow:0 inSection:section]]];
}
- (void)controller:(NSFetchedResultsController *)controller didChangeObject:(id)anObject
        atIndexPath:(NSIndexPath *)indexPath forChangeType:(NSFetchedResultsChangeType)type
        newIndexPath:(NSIndexPath *)newIndexPath {
    switch(type) {
        case NSFetchedResultsChangeInsert:
            [_agent insertCell:newIndexPath];
            break;

        case NSFetchedResultsChangeDelete:
            [_agent deleteCell:indexPath];
            break;

        case NSFetchedResultsChangeUpdate:
            [_agent changeUpdateCell:indexPath];
            break;

        case NSFetchedResultsChangeMove:
            [_agent changeMoveCell:indexPath toIndexPath:newIndexPath];
            break;
    }
}
@end

File: lib/ViewObject/FRCAgentViewObject.swift
//
//  FRCAgentViewObject.swift
//  TableViewAgent
//
//  Created by akuraru on 2014/07/31.
//  Copyright (c) 2014年 P.I.akura. All rights reserved.
//

import Foundation
import CoreData

class Support :NSObject, NSFetchedResultsControllerDelegate {
    weak var agent :TableViewAgent<NSObject, NSObject>!
    init(agent :TableViewAgent<NSObject, NSObject>) {
        self.agent = agent
    }
    func controller(controller: NSFetchedResultsController!, didChangeObject anObject: AnyObject!, atIndexPath indexPath: NSIndexPath!, forChangeType type: NSFetchedResultsChangeType, newIndexPath: NSIndexPath!) {
        switch(type) {
        case .Insert:
            agent.insertCell(newIndexPath)
        case .Delete:
            agent.deleteCell(indexPath)
        case .Update:
            agent.changeUpdateCell(indexPath)
        case .Move:
            agent.changeMoveCell(indexPath, newIndexPath: newIndexPath)
        }
    }
}

class FRCAgentViewObject<T: NSObject, U: NSObject>: AgentViewObject<T, U> {
    let controller :NSFetchedResultsController!
    weak var agent :TableViewAgent<T, U>!
    let support:Support!
    
    init(controller :NSFetchedResultsController, agent: TableViewAgent<T, U>, transform: T -> U = ({t in t} as T -> U)) {
        super.init(transform)
        self.agent = agent
        self.controller = controller
        self.support = Support(agent: a())
        controller.delegate = self.support
    }
    func a() -> TableViewAgent<NSObject, NSObject> {
        return agent as Any as TableViewAgent<NSObject, NSObject>
    }
    override func sectionCount() -> Int {
        if let s = controller.sections {
            if s.count == 1 && s[0].numberOfObjects == 0 {
                return 0
            }
            return s.count
        } else {
            return 0
        }
    }
    override func countInSection(section :Int) -> Int {
        if let s = controller.sections {
            return s[section].numberOfObjects
        } else {
            return 0;
        }
    }
    override func objectAtIndexPath(indexPath :NSIndexPath) -> U {
        return transform(controller.objectAtIndexPath(indexPath) as T)
    }
    override func removeObjectAtIndexPath(indexPath :NSIndexPath) {
    }
    override func existObject(indexPath :NSIndexPath) -> Bool {
        return indexPath.section < sectionCount() && indexPath.row < countInSection(indexPath.section)
    }
    override func sectionObjects(section :Int) -> U {
        return transform(objectAtIndexPath(NSIndexPath(forRow: 0, inSection: section)) as T)
    }
    
    override func changeObject(object :T) { }
    override func addObject(object :T,inSection section :Int) { }
}
